### ISSUE TITLE
PermutedDimsArray constant-propagation / inference

### ISSUE THREAD

**Issue by kcajf** [JuliaLang/julia#33435]
```
julia> a = rand(2, 2)
2×2 Array{Float64,2}:
 0.159947  0.956223
 0.712618  0.864652

julia> @code_warntype (x -> PermutedDimsArray(x, (2, 1)))(a)
Variables
  #self#::Core.Compiler.Const(var"##28#29"(), false)
  x::Array{Float64,2}

Body::PermutedDimsArray{Float64,2,_A,_B,Array{Float64,2}} where _B where _A
1 ─ %1 = Core.tuple(2, 1)::Core.Compiler.Const((2, 1), false)
│   %2 = Main.PermutedDimsArray(x, %1)::PermutedDimsArray{Float64,2,_A,_B,Array{Float64,2}} where _B where _A
└──      return %2
```
I feel like this should be inferrable, no?

**Comment by jipolanco** [comment]
The problem is that the permutation is encoded as a parameter of the `PermutedDimsArray` type.

One workaround would be to add an alternative `PermutedDimsArray` constructor taking a compile-time permutation, for instance `perm = Val((2, 1))`. In this case, the `_A` parameter in your example would be inferred by the compiler.

The `_B` parameter (the inverse permutation of `perm`) can also be evaluated at compile-time from `perm`. A possible implementation, using generated functions, is available [here](https://github.com/jipolanco/PencilFFTs.jl/blob/e8300154acb9d274b79cf249013b0984cdae2b5e/src/Pencils/permutations.jl#L35) (this actually computes the relative permutation between two permutations, which is a bit more general, but it's basically the same).

I'd be glad to submit a PR if there is interest in this kind of solution.

**Comment by kcajf** [comment]
True, though in the case I gave the permutation is a compile time constant, so it seems reasonable that it would be inferred without using a Val type

**Comment by adienes** [comment]
infers since 1.12
